pymedphys/_gui/streamlit/pseudonymise.py

```diff
@@ -0,0 +1,297 @@
+    """
+
+    zip_path_list: List = list()
+    if file_buffer_list is not None and len(file_buffer_list) > 0:
+        my_date_time = datetime.datetime.now()
+        str_now_datetime = my_date_time.strftime("%Y%m%d_%H%M%S")
+        zipfile_basename = f"Pseudonymised_{str_now_datetime}"
+
+        bad_data = False
+        index_to_fifty_mbyte_increment = _gen_index_list_to_fifty_mbyte_increment(
+            file_buffer_list
+        )
+
+        st.write(index_to_fifty_mbyte_increment)
+
+        zip_count = 0
+        start_index = 0
+        for end_index in index_to_fifty_mbyte_increment:
+            if start_index == end_index:
+                break
+            zip_count += 1
+            zipfile_name = f"{zipfile_basename}.{zip_count}.zip"
+            zipfile_path = pjoin(DOWNLOADS_PATH, zipfile_name)
+            zip_bytes_io = io.BytesIO()
+            bad_data = _zip_pseudo_fifty_mbytes(
+                file_buffer_list[start_index:end_index],
+                zipfile_path,
+                zip_bytes_io=zip_bytes_io,
+            )
+            start_index = end_index
+            if bad_data:
+                if zip_bytes_io is not None:
+                    zip_bytes_io.close()
+                    del zip_bytes_io
+                else:
+                    remove_file(zipfile_name)
+                st.text("Problem processing DICOM data")
+                return list()
+            else:
+                if zip_bytes_io is not None:
+                    link_to_zipbuffer_download(zipfile_name, zip_bytes_io.getvalue())
+                    zip_bytes_io.close()
+                    del zip_bytes_io
+                else:
+                    link_to_zip_download(pathlib.Path(zipfile_path))
+                    zip_path_list.append(zipfile_path)
+
+    return zip_path_list
+
+
+def _gen_index_list_to_fifty_mbyte_increment(file_buffer_list):
+    """Create a list of indexes marking the ~50 MByte increments of cumulative size
+    of the file_buffers provided by the streamlit.file_uploader
+
+    Parameters
+    ----------
+    file_buffer_list : ``list``
+        list of file_buffers (derived from io.BytesIO) obtained from streamlit.file_uploader
+
+    Returns
+    -------
+    ``list``
+        list of indexes marking the ~50 MByte increments, including the last index
+    """
+
+    file_count = 0
+    index_to_fifty_mbyte_increment = list()
+    size = 0
+    for uploaded_file_buffer in file_buffer_list:
+        file_count += 1
+        size += uploaded_file_buffer.size
+        if size > 50000000:
+            index_to_fifty_mbyte_increment.append(file_count)
+            size = 0
+
+    if size != 0:
+        index_to_fifty_mbyte_increment.append(file_count)
+
+    return index_to_fifty_mbyte_increment
+
+
+st.set_option("deprecation.showfileUploaderEncoding", False)
+
+uploaded_file_buffer_list = st.file_uploader(
+    "Files to pseudonymise, refresh page after downloading zip(s)",
```

Comment:
From a user interface stand point, I believe it is possible to trigger that download automatically once the file is ready, and then, once it has been downloaded a rerun of the streamlit application can be automatically triggered without needing to have the user refresh the page themselves.

Reply:
These are both things I wanted to do, but didn't find the programmable refresh/rerun (that you contributed to streamlit) until after creating the PR.  So, yes, I want to address this, I think it would provide a significant improvement to the UX.
I'm not sure how one triggers the download automatically.

Reply:
Regarding the automatic download, does the following do the trick?

https://stackoverflow.com/a/156715

Worth also putting in display:None as in the following comment:

https://stackoverflow.com/questions/156686/how-to-start-automatic-download-of-a-file-in-internet-explorer/156715#comment27313065_156715

Given its always going to be a zip here, that should do the trick. Haven't tested it though.

Reply:
Better than the IFrame option actually, would be to open the zip URL in a new tab using `window.open`

https://www.codexworld.com/how-to/open-url-in-new-browser-window-tab-javascript/

That way you can refresh the page and it won't be a problem if the zip hasn't finished its "start download" process as it will begin its downloading in a new tab (which should then promptly disappear).

Reply:
If you'd like, I can handle all that in the custom component PR, separate to this PR.

Reply:
> If you'd like, I can handle all that in the custom component PR, separate to this PR.
Yes, please, with sugar on top :-)